source-batchsql/main.go

```diff
@@ -0,0 +1,480 @@
+	boilerplate "github.com/estuary/connectors/source-boilerplate"
+	pc "github.com/estuary/flow/go/protocols/capture"
+	pf "github.com/estuary/flow/go/protocols/flow"
+	"github.com/invopop/jsonschema"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+
+	_ "github.com/jackc/pgx/v4/stdlib"
+)
+
+// Config tells the connector how to connect to and interact with the source database.
+type Config struct {
+	Address  string         `json:"address" jsonschema:"title=Server Address,description=The host or host:port at which the database can be reached." jsonschema_extras:"order=0"`
+	User     string         `json:"user" jsonschema:"default=flow_capture,description=The database user to authenticate as." jsonschema_extras:"order=1"`
+	Password string         `json:"password" jsonschema:"description=Password for the specified database user." jsonschema_extras:"secret=true,order=2"`
+	Database string         `json:"database" jsonschema:"default=postgres,description=Logical database name to capture from." jsonschema_extras:"order=3"`
+	Advanced advancedConfig `json:"advanced,omitempty" jsonschema:"title=Advanced Options,description=Options for advanced users. You should not typically need to modify these." jsonschema_extra:"advanced=true"`
+	// TODO(wgd): Add network tunnel support
+}
+
+type advancedConfig struct {
+	SSLMode string `json:"sslmode,omitempty" jsonschema:"title=SSL Mode,description=Overrides SSL connection behavior by setting the 'sslmode' parameter.,enum=disable,enum=allow,enum=prefer,enum=require,enum=verify-ca,enum=verify-full"`
+}
+
+// Validate checks that the configuration possesses all required properties.
+func (c *Config) Validate() error {
+	var requiredProperties = [][]string{
+		{"address", c.Address},
+		{"user", c.User},
+		{"password", c.Password},
+	}
+	for _, req := range requiredProperties {
+		if req[1] == "" {
+			return fmt.Errorf("missing '%s'", req[0])
+		}
+	}
+	return nil
+}
+
+// SetDefaults fills in the default values for unset optional parameters.
+func (c *Config) SetDefaults() {
+	// The address config property should accept a host or host:port
+	// value, and if the port is unspecified it should be the PostgreSQL
+	// default 5432.
+	if !strings.Contains(c.Address, ":") {
+		c.Address += ":5432"
+	}
+}
+
+// ToURI converts the Config to a DSN string.
+func (c *Config) ToURI() string {
+	var address = c.Address
+	var uri = url.URL{
+		Scheme: "postgres",
+		Host:   address,
+		User:   url.UserPassword(c.User, c.Password),
+	}
+	if c.Database != "" {
+		uri.Path = "/" + c.Database
+	}
+	var params = make(url.Values)
+	if c.Advanced.SSLMode != "" {
+		params.Set("sslmode", c.Advanced.SSLMode)
+	}
+	if len(params) > 0 {
+		uri.RawQuery = params.Encode()
+	}
+	return uri.String()
+}
+
+// Resource represents the capture configuration of a single resource binding.
+type Resource struct {
+	Name         string   `json:"name" jsonschema:"title=Name,description=The unique name of this resource."`
```

Comment:
This is a hacky way of getting some sort of "unique name" for a given capture binding. I don't think I ended up using it, but this is what persistent state would be associated with if I had implemented that. There's obviously some overlap with what we discussed yesterday about bindings having a stable ID.